### PR TITLE
[MLIR][Affine] Update ::fold() to have constant basis attr for affine.delinearize_index/linearize_index wherever applicable

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit updates `::fold()` to have constant(CST)
   attribute for affine.delinearize_index/linearize_index op's basis
   wherever applicable.
-- Essentially the code checks if the mixed basis OpFoldResult
   set contains any constant SSA value and converts it to a constant
   integer instead.

Signed-off-by: Abhishek Varma <abhvarma@amd.com>

### CODE REVIEW

**Review by krzysz00 — COMMENTED** [review]
I figure that this can be done in `fold()`, looking at whether the `getDynamicBasis()` adaptor has some non-null `Attribute`s

**Comment by Abhishek-Varma** [comment]
> I figure that this can be done in `fold()`, looking at whether the `getDynamicBasis()` adaptor has some non-null `Attribute`s

Hi @krzysz00 - I've made the changes. I figured I've to use `MutableOperandRange` for updating the dynamic basis - hence wasn't able to work with `::fold` earlier.

Please take a look. Thanks!

**Review by krzysz00 — COMMENTED** [review]
Could you double-check the requirements on `fold()`. I have the vague sense that if you mutated the operation you need to return the results of the operation you just mutated.

(Also, it might be a better move to look through the dynamic basis for constants and, if you find any non-null `Attribute`s, to go swap them in at the right spot)

**Comment by Abhishek-Varma** [comment]
> Could you double-check the requirements on `fold()`. I have the vague sense that if you mutated the operation you need to return the results of the operation you just mutated.
> 
Hi @krzysz00 - I've made the changes.
For `affine.delinearize_index` I'm returning `success()` whereas for `affine.linearize_index` I'm returning the result of the operation.

> (Also, it might be a better move to look through the dynamic basis for constants and, if you find any non-null `Attribute`s, to go swap them in at the right spot)

Correct, but swapping them at the right spot would need `mixedBasis` (or `staticBasis`) too - so I'm going ahead with inferring the constants from `mixedBasis` itself. That should be okay to go ahead with.

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 1]
```diff
@@ -4789,6 +4829,15 @@ LogicalResult AffineLinearizeIndexOp::verify() {
...
 
 OpFoldResult AffineLinearizeIndexOp::fold(FoldAdaptor adaptor) {
+  std::optional<SmallVector<int64_t>> maybeStaticBasis =
+      foldCstValueToCstAttrBasis(getMixedBasis(), getDynamicBasisMutable(),
+                                 adaptor.getDynamicBasis());
+  if (maybeStaticBasis) {
+    setStaticBasis(*maybeStaticBasis);
+    return getResult();
+  }
+  // setStaticBasis(foldCstValueToCstAttrBasis(
```
Comment (krzysz00) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4839]: Nit: commented code

**Review by krzysz00 — APPROVED** [review]
One minor note, otherwise, approved, thank you!

**Pull request merged**